Panic when searching empty record (regression at commit 9bd1d59)

Looks like we hit a regression starting with `zq` commit `9bd1d59` which is associated with #1511 (cc: @nwt).

As of `zq` commit `763a0a1` when #1069 was verified, it was possible to both read and search empty records like the following:

```
$ zq -version
Version: v0.20.0-22-g763a0a19

$ echo '{"ja3s":{}}' | zq -t -
#0:record[ja3s:record[]]
0:[[]]

$ echo '{"ja3s":{}}' | zq -t 'foo' -
[no output, but that's expected since the search was unsuccessful]
```

However, starting with `zq` commit `9bd1d59` which is associated with #1511, now the attempted search causes a panic.

```
$ zq -version
Version: v0.22.0-71-g9bd1d594

$ echo '{"ja3s":{}}' | zq -t -
#0:record[ja3s:record[]]
0:[[]]

$ echo '{"ja3s":{}}' | zq -t 'foo' -
panic: runtime error: index out of range [0] with length 0
goroutine 41 [running]:
runtime/debug.Stack(0x1054774, 0x9, 0xc000038100)
	/usr/local/opt/go/libexec/src/runtime/debug/stack.go:24 +0x9f
github.com/brimsec/zq/zqe.RecoverError(0x1830040, 0xc000036180, 0xc000036180, 0x8)
	/Users/phil/work/zq/zqe/error.go:153 +0x26
github.com/brimsec/zq/driver.(*mux).safeGet.func1(0xc0005b9f70)
	/Users/phil/work/zq/driver/mux.go:50 +0x48
panic(0x1830040, 0xc000036180)
	/usr/local/opt/go/libexec/src/runtime/panic.go:969 +0x1b9
github.com/brimsec/zq/filter.(*fieldNameIter).next(0xc0005a6060, 0x0, 0x0, 0x0)
	/Users/phil/work/zq/filter/fieldnamefinder.go:95 +0x40d
github.com/brimsec/zq/filter.searchRecordString.func2(0xc000138500, 0xc000138500)
	/Users/phil/work/zq/filter/filter.go:205 +0x371
github.com/brimsec/zq/scanner.(*scanner).Read(0xc0005aa060, 0x0, 0x64, 0xc000120000)
	/Users/phil/work/zq/scanner/scanner.go:98 +0x1db
github.com/brimsec/zq/zbuf.ReadBatch(0x19c3040, 0xc0005aa060, 0x64, 0x0, 0x0, 0x0, 0x0)
	/Users/phil/work/zq/zbuf/batch.go:37 +0xa3
github.com/brimsec/zq/scanner.(*scanner).Pull(0xc0005aa060, 0x0, 0x0, 0x0, 0x0)
	/Users/phil/work/zq/scanner/scanner.go:73 +0x45
github.com/brimsec/zq/driver.(*namedScanner).Pull(0xc0005a2040, 0x0, 0x0, 0x0, 0x0)
	/Users/phil/work/zq/driver/compile.go:63 +0x3a
github.com/brimsec/zq/proc/pass.(*Proc).Pull(0xc0005a8070, 0x0, 0x0, 0x0, 0x0)
	/Users/phil/work/zq/proc/pass/pass.go:17 +0x33
github.com/brimsec/zq/driver.(*mux).safeGet(0xc0005a2060, 0x0, 0x0, 0x0, 0x0)
	/Users/phil/work/zq/driver/mux.go:52 +0x6c
github.com/brimsec/zq/driver.(*mux).run(0xc0005a2060)
	/Users/phil/work/zq/driver/mux.go:65 +0x32
created by github.com/brimsec/zq/driver.(*muxOutput).Pull.func1
	/Users/phil/work/zq/driver/mux.go:103 +0x5d
```

--- a/filter/fieldnamefinder.go
+++ b/filter/fieldnamefinder.go
@@ -81,7 +81,10 @@ type fieldNameIterInfo struct {
 
 func (f *fieldNameIter) init(t *zng.TypeRecord) {
 	f.buf = f.buf[:0]
-	f.stack = append(f.stack[:0], fieldNameIterInfo{t.Columns, 0})
+	f.stack = f.stack[:0]
+	if len(t.Columns) > 0 {
+		f.stack = append(f.stack, fieldNameIterInfo{t.Columns, 0})
+	}
 }
 
 func (f *fieldNameIter) done() bool {
@@ -89,13 +92,13 @@ func (f *fieldNameIter) done() bool {
 }
 
 func (f *fieldNameIter) next() []byte {
-	// Step into records.
+	// Step into non-empty records.
 	for {
 		info := &f.stack[len(f.stack)-1]
 		col := info.columns[info.offset]
 		f.buf = append(f.buf, "."+col.Name...)
 		t, ok := zng.AliasedType(col.Type).(*zng.TypeRecord)
-		if !ok {
+		if !ok || len(t.Columns) == 0 {
 			break
 		}
 		f.stack = append(f.stack, fieldNameIterInfo{t.Columns, 0})

--- a/filter/fieldnamefinder_test.go
+++ b/filter/fieldnamefinder_test.go
@@ -9,11 +9,12 @@ import (
 )
 
 func TestFieldNameIter(t *testing.T) {
-	const typeString = "record[r1:record[r2:record[s:string,r3:record[t:time]],a:array[int64],r4:record[i:ip]]]"
+	const typeString = "record[r1:record[r2:record[s:string,r3:record[t:time]],a:array[int64],r4:record[i:ip]],empty:record[]]"
 	typ, err := resolver.NewContext().LookupByName(typeString)
 	require.NoError(t, err)
 	var f fieldNameIter
 	f.init(typ.(*zng.TypeRecord))
+	require.False(t, f.done())
 	require.Exactly(t, "r1.r2.s", string(f.next()))
 	require.False(t, f.done())
 	require.Exactly(t, "r1.r2.r3.t", string(f.next()))
@@ -21,5 +22,14 @@ func TestFieldNameIter(t *testing.T) {
 	require.Exactly(t, "r1.a", string(f.next()))
 	require.False(t, f.done())
 	require.Exactly(t, "r1.r4.i", string(f.next()))
+	require.False(t, f.done())
+	require.Exactly(t, "empty", string(f.next()))
+	require.True(t, f.done())
+}
+func TestFieldNameIterEmptyTopLevelRecord(t *testing.T) {
+	typ, err := resolver.NewContext().LookupByName("record[]")
+	require.NoError(t, err)
+	var f fieldNameIter
+	f.init(typ.(*zng.TypeRecord))
 	require.True(t, f.done())
 }

--- a/filter/filter_test.go
+++ b/filter/filter_test.go
@@ -499,6 +499,22 @@ func TestFilters(t *testing.T) {
 		{"rec.str", true},
 	})
 
+	// Test searching an empty top-level record
+	tzng = `
+#0:record[]
+0:[]`
+	runCases(t, tzng, []testcase{
+		{"empty", false},
+	})
+
+	// Test searching an empty nested record
+	tzng = `
+#0:record[empty:record[]]
+0:[[]]`
+	runCases(t, tzng, []testcase{
+		{"empty", true},
+	})
+
 }
 
 func TestBadFilter(t *testing.T) {

